rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,451 @@
+                        self.protocol_version = conn.protocol_version;
+                    }
+                    let tx = self.new_tx(msg);
+                    self.transactions.push(tx);
+                }
+                Err(nom::Err::Incomplete(v)) => {
+                    if let nom::Needed::Size(s) = v {
+                        SCLogDebug!("incomplete request: needed {}", s);
+                    }
+                    self.request_buffer.extend_from_slice(current);
+                    break;
+                }
+                Err(_) => {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+
+    fn parse_response(&mut self, input: &[u8]) -> bool {
+        if input.len() == 0 {
+            return true;
+        }
+
+        self.response_buffer.extend(input);
+        let tmp: Vec<u8>;
+        let mut current = {
+            tmp = self.response_buffer.split_off(0);
+            tmp.as_slice()
+        };
+
+        while current.len() > 0 {
+            SCLogDebug!("response: handling {}", current.len());
+            match parse_message(current, self.protocol_version) {
+                Ok((rem, msg)) => {
+                    SCLogDebug!("replacing current ({}) with rem ({})", current.len(), rem.len());
+                    current = rem;
+                    let tx = self.new_tx(msg);
+                    self.transactions.push(tx);
+                }
+                Err(nom::Err::Incomplete(v)) => {
+                    if let nom::Needed::Size(s) = v {
+                        SCLogDebug!("incomplete response: needed {}", s);
+                    }
+                    self.response_buffer.extend_from_slice(current);
+                    break;
+                }
+                Err(_) => {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+
+    fn tx_iterator(
+        &mut self,
+        min_tx_id: u64,
+        state: &mut u64,
+    ) -> Option<(&MQTTTransaction, u64, bool)> {
+        let mut index = *state as usize;
+        let len = self.transactions.len();
+
+        while index < len {
+            let tx = &self.transactions[index];
+            if tx.tx_id < min_tx_id + 1 {
+                index += 1;
+                continue;
+            }
+            *state = index as u64;
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
+        }
+
+        return None;
+    }
+}
+
+// C exports.
+
+export_tx_get_detect_state!(rs_mqtt_tx_get_detect_state, MQTTTransaction);
+export_tx_set_detect_state!(rs_mqtt_tx_set_detect_state, MQTTTransaction);
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_probing_parser(
+    _flow: *const Flow,
+    _direction: u8,
+    input: *const u8,
+    input_len: u32,
+    _rdir: *mut u8,
+) -> AppProto {
+    let buf = build_slice!(input, input_len as usize);
+    if parse_message(buf, 3).is_ok() {
```

Comment:
Why magic 3 ?

Reply:
This is the default protocol number that is assumed and gets overwritten once we have determined the actual version from the connect handshake. In v2, I will use a Rust constant to clarify that to the reader of the code.

Reply:
Thanks